File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root: return ""
        s = ""
        q = deque()
        q.append(root)
        while len(q) != 0:
            cur = q.popleft()
            if cur:
                s += str(cur.val) + "#"
                q.append(cur.left)
                q.append(cur.right)
            else:
                s += "n#"
        return s

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == "":
            return None
        vals = data.split("#")[:-1]       
        root = TreeNode(int(vals[0]))
        q = deque()
        q.append(root)
        i = 1
        n = len(vals)
        while len(q) != 0 and i < n:
            cur = q.popleft()
            if i < n and vals[i] != "n":
                cur.left = TreeNode(int(vals[i]))
                q.append(cur.left)
            else:
                cur.left = None
            i += 1
            if i < n and vals[i] != "n":
                cur.right = TreeNode(int(vals[i]))
                q.append(cur.right)
            else:
                cur.right = None
            i += 1
        return root

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root)
            return "";
        string s;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()){
            TreeNode* cur = q.front();
            q.pop();
            if (cur){
                s += to_string(cur->val) + "#";
                q.push(cur->left);
                q.push(cur->right);
            } else 
                s += "n#";
        }
        return s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.empty())
            return nullptr;
        vector<string> tokens;
        string token;
        for (char ch : data){
            if (ch == '#'){
                tokens.push_back(token);
                token.clear();
            } else
                token.push_back(ch);
        }
        TreeNode* root = new TreeNode(stoi(tokens[0]));
        queue<TreeNode*> q;
        q.push(root);
        int i = 1;
        int n = (int)tokens.size();
        while (i < n && !q.empty()){
            TreeNode* cur = q.front();
            q.pop();
            if (i < n && tokens[i] != "n"){
                cur->left = new TreeNode(stoi(tokens[i]));
                q.push(cur->left);
            }
            i++;
            if (i < n && tokens[i] != "n"){
                cur->right = new TreeNode(stoi(tokens[i]));
                q.push(cur->right);
            }
            i++;
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));